larpandoracontent/LArTrackShowerId/TrackShowerIdFeatureTool.cc

```diff
@@ -568,34 +591,41 @@ ThreeDPCAFeatureTool::ThreeDPCAFeatureTool()
 void ThreeDPCAFeatureTool::Run(SupportVectorMachine::DoubleVector &featureVector, const Algorithm *const pAlgorithm,
     const pandora::ParticleFlowObject *const pInputPfo)
 {
-     if (PandoraContentApi::GetSettings(*pAlgorithm)->ShouldDisplayAlgorithmInfo())
+	if (PandoraContentApi::GetSettings(*pAlgorithm)->ShouldDisplayAlgorithmInfo())
         std::cout << "----> Running Algorithm Tool: " << this->GetInstanceName() << ", " << this->GetType() << std::endl;
 
-    // Need the 3D cluster and hits to calculate PCA components
-    ClusterList threeDClusterList;
-    LArPfoHelper::GetThreeDClusterList(pInputPfo, threeDClusterList);
-
-    if (threeDClusterList.empty())
-        return;
+	// Need the 3D cluster and hits to calculate PCA components
+	ClusterList threeDClusterList;
+	LArPfoHelper::GetThreeDClusterList(pInputPfo, threeDClusterList);
 
-    CaloHitList threeDCaloHitList;
-    LArPfoHelper::GetCaloHits(pInputPfo, TPC_3D, threeDCaloHitList);
+	if (threeDClusterList.empty())
+		return;
 
-    if (threeDCaloHitList.empty())
-        return;
+	CaloHitList threeDCaloHitList;
+	LArPfoHelper::GetCaloHits(pInputPfo, TPC_3D, threeDCaloHitList);
 
-    // Run the PCA analysis
-    CartesianVector centroid(0.f, 0.f, 0.f);
-    LArPcaHelper::EigenVectors eigenVecs;
-    LArPcaHelper::EigenValues eigenValues(0.f, 0.f, 0.f);
-    LArPcaHelper::RunPca(threeDCaloHitList, centroid, eigenValues, eigenVecs);
-    const float principalEigenvalue(eigenValues.GetX()), secondaryEigenvalue(eigenValues.GetY()), tertiaryEigenvalue(eigenValues.GetZ());
+	if (threeDCaloHitList.empty())
+		return;
```

Comment:
What happens if you return and then the feature vector doesn't have the expected size? Do we return anywhere else, in any of the other tools, rather than pushing back a null/dummy value into the feature vector? Won't this cause significant problems and mixing up of doubles as feature values?